File: src/core/controller.ts
/**
 * NES 控制器模擬
 * 
 * 標準 NES 控制器有 8 個按鈕，按讀取順序：
 * A, B, Select, Start, Up, Down, Left, Right
 */

/**
 * 控制器按鈕
 */
export enum ControllerButton {
  A = 0,
  B = 1,
  Select = 2,
  Start = 3,
  Up = 4,
  Down = 5,
  Left = 6,
  Right = 7,
}

/**
 * 控制器按鈕位元遮罩
 */
export const BUTTON_MASKS = {
  [ControllerButton.A]: 1 << 7,
  [ControllerButton.B]: 1 << 6,
  [ControllerButton.Select]: 1 << 5,
  [ControllerButton.Start]: 1 << 4,
  [ControllerButton.Up]: 1 << 3,
  [ControllerButton.Down]: 1 << 2,
  [ControllerButton.Left]: 1 << 1,
  [ControllerButton.Right]: 1 << 0,
};

/**
 * 控制器類別
 */
export class Controller {
  /** 按鈕狀態 (8 位元，每個位元代表一個按鈕) */
  private buttonState: number = 0;

  constructor() {
    this.buttonState = 0;
  }

  /**
   * 設定按鈕狀態
   * @param button 按鈕
   * @param pressed 是否按下
   */
  public setButton(button: ControllerButton, pressed: boolean): void {
    const mask = BUTTON_MASKS[button];
    if (pressed) {
      this.buttonState |= mask;
    } else {
      this.buttonState &= ~mask;
    }
  }

  /**
   * 取得按鈕狀態
   * @param button 按鈕
   * @returns 是否按下
   */
  public isButtonPressed(button: ControllerButton): boolean {
    return (this.buttonState & BUTTON_MASKS[button]) !== 0;
  }

  /**
   * 取得完整的按鈕狀態 (用於序列化讀取)
   * @returns 8 位元狀態值
   */
  public getState(): number {
    return this.buttonState;
  }

  /**
   * 重置所有按鈕
   */
  public reset(): void {
    this.buttonState = 0;
  }
}

/**
 * 預設鍵盤映射 (玩家 1)
 */
export const DEFAULT_KEYBOARD_MAP_P1: Record<string, ControllerButton> = {
  'KeyZ': ControllerButton.A,
  'KeyX': ControllerButton.B,
  'ShiftRight': ControllerButton.Select,
  'Enter': ControllerButton.Start,
  'ArrowUp': ControllerButton.Up,
  'ArrowDown': ControllerButton.Down,
  'ArrowLeft': ControllerButton.Left,
  'ArrowRight': ControllerButton.Right,
};

/**
 * 預設鍵盤映射 (玩家 2)
 */
export const DEFAULT_KEYBOARD_MAP_P2: Record<string, ControllerButton> = {
  'KeyN': ControllerButton.A,
  'KeyM': ControllerButton.B,
  'KeyQ': ControllerButton.Select,
  'KeyW': ControllerButton.Start,
  'KeyI': ControllerButton.Up,
  'KeyK': ControllerButton.Down,
  'KeyJ': ControllerButton.Left,
  'KeyL': ControllerButton.Right,
};

/**
 * 鍵盤輸入處理器
 */
export class KeyboardInputHandler {
  private controller: Controller;
  private keyMap: Record<string, ControllerButton>;

  constructor(controller: Controller, keyMap: Record<string, ControllerButton>) {
    this.controller = controller;
    this.keyMap = keyMap;
  }

  /**
   * 處理按鍵按下
   */
  public handleKeyDown(event: KeyboardEvent): void {
    const button = this.keyMap[event.code];
    if (button !== undefined) {
      this.controller.setButton(button, true);
      event.preventDefault();
    }
  }

  /**
   * 處理按鍵釋放
   */
  public handleKeyUp(event: KeyboardEvent): void {
    const button = this.keyMap[event.code];
    if (button !== undefined) {
      this.controller.setButton(button, false);
      event.preventDefault();
    }
  }

  /**
   * 綁定到 window 事件
   */
  public bind(): void {
    window.addEventListener('keydown', this.handleKeyDown.bind(this));
    window.addEventListener('keyup', this.handleKeyUp.bind(this));
  }

  /**
   * 解除綁定
   */
  public unbind(): void {
    window.removeEventListener('keydown', this.handleKeyDown.bind(this));
    window.removeEventListener('keyup', this.handleKeyUp.bind(this));
  }
}

File: nes-wasm/src/snes/controller.rs
// ============================================================
// SNES 控制器 - 12 鍵手把模擬
// ============================================================
// B, Y, Select, Start, Up, Down, Left, Right, A, X, L, R
// 透過 $4016/$4017 自動讀取 ($4218-$421F)
// ============================================================

/// SNES 控制器按鈕定義
/// 按鈕在 16-bit 暫存器中的位元位置
pub mod buttons {
    pub const B: u16      = 1 << 15;  // bit 15
    pub const Y: u16      = 1 << 14;  // bit 14
    pub const SELECT: u16 = 1 << 13;  // bit 13
    pub const START: u16  = 1 << 12;  // bit 12
    pub const UP: u16     = 1 << 11;  // bit 11
    pub const DOWN: u16   = 1 << 10;  // bit 10
    pub const LEFT: u16   = 1 << 9;   // bit 9
    pub const RIGHT: u16  = 1 << 8;   // bit 8
    pub const A: u16      = 1 << 7;   // bit 7
    pub const X: u16      = 1 << 6;   // bit 6
    pub const L: u16      = 1 << 5;   // bit 5
    pub const R: u16      = 1 << 4;   // bit 4
}

pub struct Controller {
    /// 當前按鈕狀態 (16-bit, 按下=1)
    pub state: u16,
    /// 自動讀取結果 (latched)
    pub auto_read_result: u16,
}

impl Controller {
    pub fn new() -> Self {
        Controller {
            state: 0,
            auto_read_result: 0,
        }
    }

    /// 設定按鈕狀態
    /// button: 0=B, 1=Y, 2=Select, 3=Start, 4=Up, 5=Down, 6=Left, 7=Right,
    ///         8=A, 9=X, 10=L, 11=R
    pub fn set_button(&mut self, button: u8, pressed: bool) {
        let mask = match button {
            0  => buttons::B,
            1  => buttons::Y,
            2  => buttons::SELECT,
            3  => buttons::START,
            4  => buttons::UP,
            5  => buttons::DOWN,
            6  => buttons::LEFT,
            7  => buttons::RIGHT,
            8  => buttons::A,
            9  => buttons::X,
            10 => buttons::L,
            11 => buttons::R,
            _  => return,
        };
        if pressed {
            self.state |= mask;
        } else {
            self.state &= !mask;
        }
    }

    /// 執行自動讀取（VBlank 期間由硬體觸發）
    pub fn auto_read(&mut self) {
        self.auto_read_result = self.state;
    }
}
